strutil/env.go

```diff
@@ -0,0 +1,253 @@
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package strutil
+
+import (
+	"fmt"
+	"os"
+	"sort"
+	"strings"
+)
+
+// RawEnvironment is a slice of key=value strings.
+//
+// Using this type is required for low-level interactions, like invoking
+// execve. In other cases using Environment is both sufficient and safer.
+type RawEnvironment []string
+
+// String returns the readable, quoted representation of the raw environment.
+func (raw RawEnvironment) String() string {
+	return Quoted(raw)
+}
+
+// Environment is an unordered map of key=value strings.
+//
+// Environment can be manipulated with available methods and eventually
+// converted to RawEnvironment. This approach discourages operations that could
+// result in duplicate environment variable definitions from being constructed.
+type Environment struct {
+	entries map[string]string
+}
+
+// NewEnvironment returns an environment with a copy of given entries.
+func NewEnvironment(entries map[string]string) *Environment {
+	env := &Environment{entries: make(map[string]string, len(entries))}
+	for key, value := range entries {
+		env.entries[key] = value
+	}
+	return env
+}
+
+func parseEnvEntry(entry string) (string, string, error) {
+	parts := strings.SplitN(entry, "=", 2)
+	if len(parts) != 2 {
+		return "", "", fmt.Errorf("cannot parse environment entry: %q", entry)
+	}
+	key, value := parts[0], parts[1]
+	if key == "" {
+		return "", "", fmt.Errorf("environment variable name cannot be empty: %q", entry)
+	}
+	return key, value, nil
+}
+
+// ParseRawEnvironment parsers raw environment.
+//
+// This function fails if any of the provided values are not in the form of
+// key=value or if there are duplicate keys.
+func ParseRawEnvironment(raw RawEnvironment) (*Environment, error) {
+	env := &Environment{entries: make(map[string]string, len(raw))}
+	for _, entry := range raw {
+		key, value, err := parseEnvEntry(entry)
+		if err != nil {
+			return nil, err
+		}
+		if _, ok := env.entries[key]; ok {
+			return nil, fmt.Errorf("cannot overwrite earlier value of %q", key)
+		}
+		env.entries[key] = value
+	}
+	return env, nil
+}
+
+// OSEnvironment returns the environment of the calling process.
+func OSEnvironment() (*Environment, error) {
+	return ParseRawEnvironment(os.Environ())
+}
+
+// Transform programmatically replaces all keys and values.
+//
+// If multiple keys are transformed into the same key the value of the last
+// (lexicographically) key is used. If the transformed key is empty then the
+// corresponding entry is removed.
+func (env *Environment) Transform(tr func(key, value string) (string, string)) {
```

Comment:
I would try to avoid this part of the change for now, also because it's unclear to me that the splitting of responsibilities make sense.  If this lives in osutil, the list of setgid discarded stuff could live there and the logic be moved to helpers that load os env and respectively prepare it for exec.

Reply:
Done.